### PR TITLE
mds: dump SnapRealm subvolume_ino

### PR DESCRIPTION
**Pull request by mchangir**
for debugging purpose

Fixes: https://tracker.ceph.com/issues/65224
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mds/SnapRealm.h [hunk 1]
```diff
@@ -92,7 +92,7 @@ struct SnapRealm {
     return (p != s.end() && *p <= last);
   }
 
-  inodeno_t get_subvolume_ino() {
+  inodeno_t get_subvolume_ino() const {
```
Comment (vshankar) [inline src/mds/SnapRealm.h:95]: Is this required for the debug change?

Comment (mchangir) [reply src/mds/SnapRealm.h:95]: yes, its required
or
we need to do away with the `const` specifier on the SnapRealm parameter in the dump `operator<<(...)` function

Comment (vshankar) [reply src/mds/SnapRealm.h:95]: So, `cached_subvolume_ino` is modified in `check_cache()` and that variable is marked mutable. You could just use `cached_subvolume_ino` in `operator<<`, but I guess this is fine too :)

**Comment by vshankar** [comment]
jenkins test api

File: src/mds/SnapRealm.h (hunk 1 above)
Comment (mchangir) [reply src/mds/SnapRealm.h:95]: yeah ... also `cached_subvolume_ino` is kept `private`

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66035](https://tracker.ceph.com/issues/66035).

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66063](https://tracker.ceph.com/issues/66063).

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66090](https://tracker.ceph.com/issues/66090).

**Comment by vshankar** [comment]
It may sound a bit odd, but it seems like this change is causing a test run failure: https://pulpito.ceph.com/vshankar-2024-05-27_05:52:49-fs-wip-vshankar-testing-20240521.063558-debug-testing-default-smithi/7727595/

Actual issues is the s-r daemon crashing

```
 ceph version 19.0.0-3872-g38f9d900 (38f9d900c83c63799fbdbe61acc9a11b0d3554a6) squid (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x127) [0x7f680cab4ba6]
 2: (ceph::register_assert_context(ceph::common::CephContext*)+0) [0x7f680cab4dd1]
 3: (SnapClient::filter(std::set<snapid_t, std::less<snapid_t>, std::allocator<snapid_t> > const&) const+0x81) [0x5637ddc8023d]
 4: (SnapRealm::build_snap_set() const+0x323) [0x5637ddc6b971]
 5: (SnapRealm::check_cache() const+0x13e) [0x5637ddc69fa0]
 6: (operator<<(std::ostream&, SnapRealm const&)+0x1fd) [0x5637ddc6a651]
 7: (CInode::decode_snap_blob(ceph::buffer::v15_2_0::list const&)+0x271) [0x5637ddbff347]
 8: (EMetaBlob::fullbit::update_inode(MDSRank*, CInode*)+0x22c) [0x5637dd8fb87e]
 9: (EMetaBlob::replay(MDSRank*, LogSegment*, int, MDPeerUpdate*)+0x2702) [0x5637dd900e94]
 10: (EUpdate::replay(MDSRank*)+0x63) [0x5637dd90b969]
 11: (MDLog::_replay_thread()+0x1f38) [0x5637ddca448c]
 12: (MDLog::ReplayThread::entry()+0x11) [0x5637ddcad9cb]
 13: (Thread::entry_wrapper()+0x43) [0x7f680ca8372d]
 14: (Thread::_entry_func(void*)+0xd) [0x7f680ca83749]
 15: /lib64/libpthread.so.0(+0x81ca) [0x7f680b6271ca]
 16: clone()
```

I don't see this crash in any of fs suite runs in the past one month.

**Comment by vshankar** [comment]
> It may sound a bit odd, but it seems like this change is causing a test run failure: https://pulpito.ceph.com/vshankar-2024-05-27_05:52:49-fs-wip-vshankar-testing-20240521.063558-debug-testing-default-smithi/7727595/
> 
> Actual issues is the s-r daemon crashing
> 
> ```
>  ceph version 19.0.0-3872-g38f9d900 (38f9d900c83c63799fbdbe61acc9a11b0d3554a6) squid (dev)
>  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x127) [0x7f680cab4ba6]
>  2: (ceph::register_assert_context(ceph::common::CephContext*)+0) [0x7f680cab4dd1]
>  3: (SnapClient::filter(std::set<snapid_t, std::less<snapid_t>, std::allocator<snapid_t> > const&) const+0x81) [0x5637ddc8023d]
>  4: (SnapRealm::build_snap_set() const+0x323) [0x5637ddc6b971]
>  5: (SnapRealm::check_cache() const+0x13e) [0x5637ddc69fa0]
>  6: (operator<<(std::ostream&, SnapRealm const&)+0x1fd) [0x5637ddc6a651]
>  7: (CInode::decode_snap_blob(ceph::buffer::v15_2_0::list const&)+0x271) [0x5637ddbff347]
>  8: (EMetaBlob::fullbit::update_inode(MDSRank*, CInode*)+0x22c) [0x5637dd8fb87e]
>  9: (EMetaBlob::replay(MDSRank*, LogSegment*, int, MDPeerUpdate*)+0x2702) [0x5637dd900e94]
>  10: (EUpdate::replay(MDSRank*)+0x63) [0x5637dd90b969]
>  11: (MDLog::_replay_thread()+0x1f38) [0x5637ddca448c]
>  12: (MDLog::ReplayThread::entry()+0x11) [0x5637ddcad9cb]
>  13: (Thread::entry_wrapper()+0x43) [0x7f680ca8372d]
>  14: (Thread::_entry_func(void*)+0xd) [0x7f680ca83749]
>  15: /lib64/libpthread.so.0(+0x81ca) [0x7f680b6271ca]
>  16: clone()
> ```
> 
> I don't see this crash in any of fs suite runs in the past one month.

This is a sample fs:workload run without this change: https://pulpito.ceph.com/vshankar-2024-05-28_17:28:53-fs:workload-wip-vshankar-testing-20240528.122300-debug-testing-default-smithi/

No MDS crashes, so, somehow this change should be the cause for the crash.

**Review by vshankar — CHANGES_REQUESTED** [review]
@mchangir PTAL at the failure.

**Pull request closed without merging**